share/vss/vss.go

```diff
@@ -0,0 +1,579 @@
+// the verifier) or because the verifier already received a Deal, it does not
+// return a complaint.
+// ProcessDeal returns an error in any case to let the user know what was the
+// error.
+// XXX API question: return value.For the moment, the default is to let the user
+// know everything that is wrong through the error.
+func (v *Verifier) ProcessDeal(d *Deal) (*Approval, *Complaint, error) {
+	if d.SecShare.I != v.index {
+		return nil, nil, errors.New("verifier: wrong index from deal")
+	}
+
+	t := int(d.T)
+
+	sid, err := sessionID(v.suite, v.dealer, v.verifiers, d.Commitments, t)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if v.aggregator == nil {
+		v.aggregator = newAggregator(v.suite, v.dealer, v.verifiers, d.Commitments, t, d.SessionID)
+	}
+
+	if err := v.verifyDeal(d, true); err != nil {
+		if err == errDealAlreadyProcessed {
+			return nil, nil, err
+		}
+		c := &Complaint{
+			SessionID: sid,
+			Index:     uint32(v.index),
+			Deal:      d,
+		}
+		var err2 error
+		if c.Signature, err2 = sign.Schnorr(v.suite, v.long, msgComplaint(c)); err2 != nil {
+			return nil, nil, err2
+		}
+
+		if err2 = v.aggregator.addComplaint(c); err2 != nil {
+			return nil, nil, err2
+		}
+		return nil, c, err
+	}
+	ap := &Approval{
+		Index:     uint32(v.index),
+		SessionID: v.sid,
+	}
+	if ap.Signature, err = sign.Schnorr(v.suite, v.long, msgApproval(ap)); err != nil {
+		return nil, nil, err
+	}
+	return ap, nil, v.aggregator.addApproval(ap)
+}
+
+// Share returns the private share that this verifier has received. It returns
+// nil if the deal is not certified or there is not enough approvals.
+func (v *Verifier) Share() *share.PriShare {
+	if !v.EnoughApprovals() || !v.DealCertified() {
+		return nil
+	}
+	return v.deal.SecShare
+}
+
+// ProcessComplaint takes a Complaint and stores it if the Complaint is valid.
+// If the verifier already saw a Complaint from the same origin, it returns an error.
+// If the Complaint is not valid, it returns an error. That does NOT mean the
+// Deal is not good, but rather only this Complaint is not valid and should not
+// be treated.  To know whether the deal is good, call first v.EnoughApproval(),
+// and if true, call v.IsDealGood().
+func (v *Verifier) ProcessComplaint(c *Complaint) error {
+	return v.verifyComplaint(c)
+}
+
+// ProcessJustification takes a DealerResponse and returns an error if
+// something went wrong during the verification. If it is the case, that
+// probably means the Dealer is acting maliciously. In order to be sure, call
+// v.EnoughApprovals() and if true, v.IsDealGood().
+// NOTE: it does *not* verify the complaint associated with since it is supposed
+// to already have received it.
+func (v *Verifier) ProcessJustification(dr *Justification) error {
+	return v.aggregator.verifyJustification(dr)
+}
+
+func (v *Verifier) ProcessApproval(ap *Approval) error {
+	return v.verifyApproval(ap)
+}
+
+type aggregator struct {
```

Comment:
Even if it's a private structure, can we have a little bit of documentation? Please?